stdlib/public/core/String.swift

```diff
@@ -981,115 +981,119 @@ extension String {
-          f
-        )
-      }
-    } else {
-      return try _foreignWithNormalizedCodeUnitsImpl(
-        outputBuffer: _castOutputBuffer(&output),
-        icuInputBuffer: _castOutputBuffer(&icuInput),
-        icuOutputBuffer: _castOutputBuffer(&icuOutput),
-        f
-      )
+  internal func _forEachUTF8(_ f: (UInt8) throws -> Void) rethrows {
+    for byte in String(_guts).utf8 {
+      try f(byte)
     }
   }
 
-  internal func _foreignWithNormalizedCodeUnitsImpl(
-    outputBuffer: UnsafeMutableBufferPointer<UInt8>,
-    icuInputBuffer: UnsafeMutableBufferPointer<UInt16>,
-    icuOutputBuffer: UnsafeMutableBufferPointer<UInt16>,
-    _ f: (UInt8) throws -> Void
-  ) rethrows {
-    var outputBuffer = outputBuffer
-    var icuInputBuffer = icuInputBuffer
-    var icuOutputBuffer = icuOutputBuffer
+  internal func _isScalarNFCQC(
+    _ scalar: Unicode.Scalar,
+    _ prevCCC: inout UInt16
+  ) -> Bool {
+    let normData: UInt16
 
-    var index = range.lowerBound
-    let cachedEndIndex = range.upperBound
+    if scalar.value < 0x300 {
+      normData = 0
+    } else {
+      normData = _swift_stdlib_getNormData(scalar.value)
+    }
 
-    var hasBufferOwnership = false
+    let ccc = normData >> 3
+    let isNFCQC = normData & 0x6 == 0
 
-    defer {
-      if hasBufferOwnership {
-        outputBuffer.deallocate()
-        icuInputBuffer.deallocate()
-        icuOutputBuffer.deallocate()
-      }
+    if prevCCC > ccc, ccc != 0 {
+      return false
     }
 
-    while index < cachedEndIndex {
-      let result = _foreignNormalize(
-        readIndex: index,
-        endIndex: cachedEndIndex,
-        guts: _guts,
-        outputBuffer: &outputBuffer,
-        icuInputBuffer: &icuInputBuffer,
-        icuOutputBuffer: &icuOutputBuffer
-      )
-      for i in 0..<result.amountFilled {
-        try f(outputBuffer[i])
-      }
-      _internalInvariant(result.nextReadPosition != index)
-      index = result.nextReadPosition
-      if result.allocatedBuffers {
-        _internalInvariant(!hasBufferOwnership)
-        hasBufferOwnership = true
-      }
+    if !isNFCQC {
+      return false
     }
+
+    prevCCC = ccc
+    return true
   }
-}
 
-internal func _fastWithNormalizedCodeUnitsImpl(
-  sourceBuffer: UnsafeBufferPointer<UInt8>,
-  outputBuffer: UnsafeMutableBufferPointer<UInt8>,
-  icuInputBuffer: UnsafeMutableBufferPointer<UInt16>,
-  icuOutputBuffer: UnsafeMutableBufferPointer<UInt16>,
-  _ f: (UInt8) throws -> Void
-) rethrows {
-  var outputBuffer = outputBuffer
-  var icuInputBuffer = icuInputBuffer
-  var icuOutputBuffer = icuOutputBuffer
-
-  var index = String.Index(_encodedOffset: 0)
-  let cachedEndIndex = String.Index(_encodedOffset: sourceBuffer.count)
-
-  var hasBufferOwnership = false
-
-  defer {
-    if hasBufferOwnership {
-      outputBuffer.deallocate()
-      icuInputBuffer.deallocate()
-      icuOutputBuffer.deallocate()
+  internal func _withNFCCodeUnits(_ f: (UInt8) throws -> Void) rethrows {
```

Comment:
Maybe "forEachNFCCodeUnit"

Reply:
We're going to want to be looking to hand-outline any code that doesn't depend on `f` that we can. Makes it easier to inline `f` into this or this into the caller (which is unfortunately sometimes necessary to inline `f`).